packages/grafana-runtime/src/services/SidecarService_EXPERIMENTAL.ts

```diff
@@ -157,3 +166,98 @@ function getMainAppPluginId() {
 
   return mainApp || 'unknown';
 }
+
+type LocalStorageHistoryOptions = {
+  storageKey: string;
+};
+
+interface LocationStorageHistory extends H.MemoryHistory {
+  getLocationObservable(): Observable<H.Location | undefined>;
+}
+
+/**
+ * Simple wrapper over the memory history that persists the location in the localStorage.
+ *
+ * @param options
+ */
+function createLocationStorageHistory(options: LocalStorageHistoryOptions): LocationStorageHistory {
+  const storedLocation = localStorage.getItem(options.storageKey);
+  const initialEntry = storedLocation ? JSON.parse(storedLocation) : '/';
+  const locationSubject = new BehaviorSubject<H.Location | undefined>(undefined);
+  const memoryHistory = H.createMemoryHistory({ initialEntries: [initialEntry] });
+
+  let currentLocation = memoryHistory.location;
+
+  function maybeUpdateLocation() {
+    if (memoryHistory.location !== currentLocation) {
+      localStorage.setItem(
+        options.storageKey,
+        JSON.stringify(pick(memoryHistory.location, 'pathname', 'search', 'hash'))
+      );
+      currentLocation = memoryHistory.location;
+      locationSubject.next(memoryHistory.location);
```

Comment:
Main change is adding this observable so we can drive rerender from a context outside of the router that uses this history. It sounds a bit weird but whether that sidecar react tree is rendered depends on whether we have anything to render in the sidecar and that depends on this location value.